File: scripts/orchestrator.py
"""
3PL Inbound Onboarding Orchestrator
=====================================
Runs all 10 automation steps in sequence.

Steps marked TODO raise NotImplementedError — the orchestrator catches them,
logs a SKIPPED warning, and continues so that implemented steps still run.

Usage:
  python orchestrator.py --input ../config/sample_onboarding_input.json
  python orchestrator.py --input input.json --yaml ../config/dev.yaml --dry-run

Steps:
  1  BTP Application Creation          [TODO - Parag Shende]
  2  BTP Value Mapping                 [TODO - Parag Shende]
  3  Solace Event Portal (5-call seq)  [TODO - Rohini Mondal]
  4  Solace Queue Subscription Patch   [TODO - Rohini Mondal]
  5  Solace Git / input.json           [TODO - Rohini Mondal + DevOps]
  6  MuleSoft Feature Branch           [TODO - Sravani Kare + DevOps]
  7  MuleSoft YAML Patch               [DONE]
  8  MuleSoft Translation Rows         [DONE]
  9  MuleSoft PR + Notify              [TODO - Sravani Kare]
  10 SharePoint / Manual Review Gate   [MANUAL]
"""
from __future__ import annotations

import argparse
import json
import sys
import textwrap
from pathlib import Path
from typing import Any, Optional

# ── path setup ───────────────────────────────────────────────────────────────
SCRIPT_DIR  = Path(__file__).parent
CONFIG_DIR  = SCRIPT_DIR.parent / "config"
OUTPUTS_DIR = SCRIPT_DIR.parent / "outputs"

sys.path.insert(0, str(SCRIPT_DIR))
from steps import (
    step1_btp_app_creation,
    step2_btp_value_mapping,
    step3_solace_event_portal,
    step4_solace_queue_patch,
    step5_solace_git_input,
    step6_mule_feature_branch,
    step9_mule_pr_notify,
)
import patch_config  # Steps 7 + 8 (DONE)


# ── ANSI colours ─────────────────────────────────────────────────────────────
class C:
    RESET  = "\033[0m"
    BOLD   = "\033[1m"
    GREEN  = "\033[92m"
    YELLOW = "\033[93m"
    RED    = "\033[91m"
    CYAN   = "\033[96m"
    DIM    = "\033[2m"


def _banner(step: int, title: str) -> None:
    print(f"\n{C.BOLD}{C.CYAN}{'='*60}{C.RESET}")
    print(f"{C.BOLD}  Step {step}: {title}{C.RESET}")
    print(f"{C.BOLD}{C.CYAN}{'='*60}{C.RESET}")


def _ok(msg: str) -> None:
    print(f"{C.GREEN}  [OK]     {msg}{C.RESET}")


def _skip(msg: str) -> None:
    print(f"{C.YELLOW}  [SKIP]   {msg}{C.RESET}")


def _warn(msg: str) -> None:
    print(f"{C.YELLOW}  [WARN]   {msg}{C.RESET}")


def _err(msg: str) -> None:
    print(f"{C.RED}  [ERROR]  {msg}{C.RESET}")


def _info(msg: str) -> None:
    print(f"{C.DIM}  [INFO]   {msg}{C.RESET}")


def _manual(msg: str) -> None:
    print(f"{C.CYAN}  [MANUAL] {msg}{C.RESET}")


# ── Orchestration state ───────────────────────────────────────────────────────

class OrchestrationState:
    """Collects outputs from each step for use in downstream steps."""

    def __init__(self) -> None:
        # Step 1 output
        self.btp_client_id: str = ""
        self.btp_app_name:   str = ""

        # Step 3 output
        self.solace_enum_version_id:  str = ""
        self.solace_event_version_id: str = ""

        # Step 6 output
        self.mule_feature_branch: str = ""

        # Step 7 output
        self.yaml_diff_summary: str = ""

        # Step 8 output
        self.translation_row_count: int = 0

        # Step-level status
        self.statuses: dict[str, str] = {}

    def record(self, step: str, status: str) -> None:
        self.statuses[step] = status

    def summary(self) -> str:
        lines = [f"\n{'='*60}", "  ORCHESTRATION SUMMARY", f"{'='*60}"]
        for step, status in self.statuses.items():
            icon = "[OK]    " if status == "ok"      else \
                   "[SKIP]  " if status == "skipped" else \
                   "[MANUAL]" if status == "manual"  else "[ERROR] "
            lines.append(f"  {icon} {step}")
        lines.append(f"{'='*60}")
        return "\n".join(lines)


# ── Step runners ─────────────────────────────────────────────────────────────

def _run_step1(data: dict, state: OrchestrationState, dry_run: bool) -> None:
    _banner(1, "BTP Application Creation")
    cfg = step1_btp_app_creation.BtpAppConfig(
        partner_id=data.get("partner_id", ""),
        # TODO: fill remaining fields from data once Parag confirms schema
    )
    try:
        if dry_run:
            _skip("dry-run: skipping Step 1")
            state.record("Step 1 — BTP App Creation", "skipped")
            return
        result = step1_btp_app_creation.run(cfg)
        state.btp_client_id = result.client_id
        state.btp_app_name  = result.app_name
        _ok(f"BTP app created — clientId: {result.client_id}")
        state.record("Step 1 — BTP App Creation", "ok")
    except NotImplementedError as e:
        _skip(str(e).splitlines()[0])
        state.record("Step 1 — BTP App Creation", "skipped")


def _run_step2(data: dict, state: OrchestrationState, dry_run: bool) -> None:
    _banner(2, "BTP Value Mapping")
    cfg = step2_btp_value_mapping.BtpValueMappingConfig(
        partner_id=state.btp_client_id or data.get("partner_id", ""),
        country_code=data.get("country_code", ""),
        # TODO: fill remaining fields once Parag confirms schema
    )
    try:
        if dry_run:
            _skip("dry-run: skipping Step 2")
            state.record("Step 2 — BTP Value Mapping", "skipped")
            return
        result = step2_btp_value_mapping.run(cfg)
        _ok("BTP value mapping deployed")
        state.record("Step 2 — BTP Value Mapping", "ok")
    except NotImplementedError as e:
        _skip(str(e).splitlines()[0])
        state.record("Step 2 — BTP Value Mapping", "skipped")


def _run_step3(data: dict, state: OrchestrationState, dry_run: bool) -> None:
    _banner(3, "Solace Event Portal Design (5-call sequence)")
    cfg = step3_solace_event_portal.SolaceEventPortalConfig(
        country_code=data.get("country_code", ""),
        region_iso=data.get("region_iso", ""),
        partner_id=state.btp_client_id or data.get("partner_id", ""),
        subscription_topics=data.get("subscription_topics", []),
        # TODO: fill api_token, enum_id, event_id, etc. from Key Vault
    )
    try:
        if dry_run:
            _skip("dry-run: skipping Step 3")
            state.record("Step 3 — Solace Event Portal", "skipped")
            return
        result = step3_solace_event_portal.run(cfg)
        state.solace_enum_version_id  = result.raw_responses.get("enum_version_id", "")
        state.solace_event_version_id = result.raw_responses.get("event_version_id", "")
        _ok("Solace Event Portal design steps complete")
        state.record("Step 3 — Solace Event Portal", "ok")
    except NotImplementedError as e:
        _skip(str(e).splitlines()[0])
        state.record("Step 3 — Solace Event Portal", "skipped")


def _run_step4(data: dict, state: OrchestrationState, dry_run: bool) -> None:
    _banner(4, "Solace Queue Subscription Patch")
    cfg = step4_solace_queue_patch.SolaceQueueConfig(
        partner_id=state.btp_client_id or data.get("partner_id", ""),
        new_subscription_topic=data.get("subscription_topic", ""),
        # TODO: fill semp_host, vpn_name_uat/prod, queue_name from Key Vault
    )
    try:
        if dry_run:
            _skip("dry-run: skipping Step 4")
            state.record("Step 4 — Solace Queue Patch", "skipped")
            return
        result = step4_solace_queue_patch.run(cfg)
        _ok("Solace queue subscription appended")
        state.record("Step 4 — Solace Queue Patch", "ok")
    except NotImplementedError as e:
        _skip(str(e).splitlines()[0])
        state.record("Step 4 — Solace Queue Patch", "skipped")


def _run_step5(data: dict, state: OrchestrationState, dry_run: bool) -> None:
    _banner(5, "Solace Git Feature Branch + input.json")
    cfg = step5_solace_git_input.SolaceGitConfig(
        country_code=data.get("country_code", ""),
        partner_id=state.btp_client_id or data.get("partner_id", ""),
        subscription_topics=data.get("subscription_topics", []),
        enum_version_id=state.solace_enum_version_id,
        event_version_id=state.solace_event_version_id,
        # TODO: fill repo_owner, repo_name, github_pat from Key Vault
    )
    try:
        if dry_run:
            _skip("dry-run: skipping Step 5")
            state.record("Step 5 — Solace Git / input.json", "skipped")
            return
        result = step5_solace_git_input.run(cfg)
        _ok(f"Solace branch + input.json committed: {result.branch_name}")
        state.record("Step 5 — Solace Git / input.json", "ok")
    except NotImplementedError as e:
        _skip(str(e).splitlines()[0])
        state.record("Step 5 — Solace Git / input.json", "skipped")


def _run_step6(data: dict, state: OrchestrationState, dry_run: bool) -> None:
    _banner(6, "MuleSoft Feature Branch Creation")
    nav = data.get("nav", {})
    cfg = step6_mule_feature_branch.MuleBranchConfig(
        partner_id=state.btp_client_id or data.get("partner_id", ""),
        country_code=data.get("country_code", data.get("country_key", "")),
        # TODO: fill repo_owner, repo_name, github_pat from Key Vault
    )
    try:
        if dry_run:
            _skip("dry-run: skipping Step 6")
            state.record("Step 6 — Mule Feature Branch", "skipped")
            return
        result = step6_mule_feature_branch.run(cfg)
        state.mule_feature_branch = result.branch_name
        _ok(f"Mule feature branch created: {result.branch_name}")
        state.record("Step 6 — Mule Feature Branch", "ok")
    except NotImplementedError as e:
        _skip(str(e).splitlines()[0])
        state.record("Step 6 — Mule Feature Branch", "skipped")


def _run_steps7_8(data: dict, yaml_path: Path, state: OrchestrationState, dry_run: bool) -> None:
    _banner(7, "MuleSoft YAML Patch + Translation Rows (Steps 7 & 8)")
    errors = patch_config.validate(data)
    if errors:
        for e in errors:
            _err(e)
        state.record("Step 7 — Mule YAML Patch", "error")
        state.record("Step 8 — Translation Rows", "error")
        return

    country_key = data["country_key"]
    if dry_run:
        _skip(f"dry-run: would patch dev.yaml with country '{country_key}'")
        state.record("Step 7 — Mule YAML Patch", "skipped")
        state.record("Step 8 — Translation Rows", "skipped")
        return

    try:
        new_block = patch_config.build_yaml_block(data)
        yaml_text = yaml_path.read_text(encoding="utf-8")
        patched   = patch_config.patch_yaml(yaml_text, country_key, new_block)
        yaml_path.write_text(patched, encoding="utf-8")
        _ok(f"dev.yaml patched with '{country_key}' block")
        state.yaml_diff_summary = new_block
        state.record("Step 7 — Mule YAML Patch", "ok")
    except ValueError as e:
        _err(str(e))
        state.record("Step 7 — Mule YAML Patch", "error")
        return

    OUTPUTS_DIR.mkdir(exist_ok=True)
    rows = patch_config.build_translation_rows(data)
    state.translation_row_count = len(rows)

    xlsx_path = OUTPUTS_DIR / f"{country_key}_translation.xlsx"
    csv_path  = OUTPUTS_DIR / f"{country_key}_translation.csv"
    patch_config.write_excel(rows, str(xlsx_path))
    patch_config.write_csv(rows, str(csv_path))
    _ok(f"Generated {len(rows)} translation rows -> {xlsx_path.name}, {csv_path.name}")
    state.record("Step 8 — Translation Rows", "ok")


def _run_step9(data: dict, state: OrchestrationState, dry_run: bool) -> None:
    _banner(9, "MuleSoft PR + Reviewer Notification")
    cfg = step9_mule_pr_notify.MulePRConfig(
        feature_branch=state.mule_feature_branch,
        country_key=data.get("country_key", ""),
        partner_id=state.btp_client_id or data.get("partner_id", ""),
        yaml_diff_summary=state.yaml_diff_summary,
        translation_row_count=state.translation_row_count,
        # TODO: fill repo_owner, repo_name, github_pat, reviewer, notification from Key Vault
    )
    try:
        if dry_run:
            _skip("dry-run: skipping Step 9")
            state.record("Step 9 — Mule PR + Notify", "skipped")
            return
        result = step9_mule_pr_notify.run(cfg)
        _ok(f"PR raised: {result.pr_url} | Notified: {result.notification_sent}")
        state.record("Step 9 — Mule PR + Notify", "ok")
    except NotImplementedError as e:
        _skip(str(e).splitlines()[0])
        state.record("Step 9 — Mule PR + Notify", "skipped")


def _run_step10(state: OrchestrationState) -> None:
    _banner(10, "SharePoint / Manual Review Gate")
    _manual("Upload translation XLSX to SharePoint TranslationData_TST list.")
    _manual("Verify DEV deployment after MuleSoft PR is merged.")
    _manual("Confirm Solace subscriptions are live in UAT.")
    state.record("Step 10 — SharePoint / Manual Review", "manual")


# ── Main ─────────────────────────────────────────────────────────────────────

def main() -> None:
    parser = argparse.ArgumentParser(
        description="3PL Inbound Onboarding Orchestrator",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=textwrap.dedent("""\
          Steps with NotImplementedError are skipped with a WARN and the run continues.
          Use --dry-run to validate input and preview outputs without writing anything.
        """),
    )
    parser.add_argument("--input", "-i", help="Path to onboarding JSON",
                        default=str(CONFIG_DIR / "sample_onboarding_input.json"))
    parser.add_argument("--yaml", "-y",  help="Path to MuleSoft dev.yaml",
                        default=str(CONFIG_DIR / "dev.yaml"))
    parser.add_argument("--dry-run", action="store_true",
                        help="Validate + preview without making any changes")
    args = parser.parse_args()

    input_path = Path(args.input)
    yaml_path  = Path(args.yaml)

    if not input_path.exists():
        _err(f"Input file not found: {input_path}")
        sys.exit(1)
    if not yaml_path.exists():
        _err(f"YAML file not found: {yaml_path}")
        sys.exit(1)

    with open(input_path, encoding="utf-8") as f:
        data = json.load(f)

    if args.dry_run:
        print(f"\n{C.YELLOW}[DRY-RUN] No files will be written or APIs called.{C.RESET}")

    print(f"\n{C.BOLD}3PL Onboarding Orchestrator{C.RESET}")
    print(f"  Input : {input_path}")
    print(f"  YAML  : {yaml_path}")
    print(f"  Output: {OUTPUTS_DIR}")

    state = OrchestrationState()

    _run_step1(data, state, args.dry_run)
    _run_step2(data, state, args.dry_run)
    _run_step3(data, state, args.dry_run)
    _run_step4(data, state, args.dry_run)
    _run_step5(data, state, args.dry_run)
    _run_step6(data, state, args.dry_run)
    _run_steps7_8(data, yaml_path, state, args.dry_run)
    _run_step9(data, state, args.dry_run)
    _run_step10(state)

    print(state.summary())


if __name__ == "__main__":
    main()

File: webapp/lib/orchestrator.ts
/**
 * 3PL Inbound Onboarding Orchestrator (TypeScript)
 * ==================================================
 * Runs all automation steps in sequence.
 *
 * Integration config (BTP, Solace, GitHub) is read from:
 *   1. input.btp / input.solace / input.github  — user-provided via the UI form
 *   2. Environment variables                     — server-level fallback for deployed instances
 *
 * Step status:
 *   1  BTP Application Creation          [STUB — Parag Shende]
 *   2  BTP Value Mapping                 [STUB — Parag Shende]
 *   3  Solace Event Portal (5-call seq)  [STUB — Rohini Mondal]
 *   4  Solace Queue Subscription Patch   [STUB — Rohini Mondal]
 *   5  Solace Git / input.json           [STUB — Rohini Mondal + DevOps]
 *   6  MuleSoft Feature Branch           [STUB — Sravani Kare + DevOps]
 *   7  MuleSoft YAML Patch               [DONE]
 *   8  MuleSoft Translation Rows         [DONE]
 *   9  MuleSoft PR + Notify              [STUB — Sravani Kare]
 *  10  SharePoint / Manual Review Gate   [MANUAL]
 */

import type { OnboardingInput, RunResult } from "./types";

import { run as runStep1 } from "./steps/step1-btp-app";
import { run as runStep2 } from "./steps/step2-btp-valuemap";
import { run as runStep3 } from "./steps/step3-solace-portal";
import { run as runStep4 } from "./steps/step4-solace-queue";
import { run as runStep5 } from "./steps/step5-solace-git";
import { run as runStep6 } from "./steps/step6-mule-branch";
import { run as runStep9 } from "./steps/step9-mule-pr";
import {
  buildYamlBlock, buildYamlBlockForEnv,
  patchYaml, patchValidCountriesList, patchNavisionInstance,
  buildTranslationRows, buildCsv,
} from "./core";
import { ensureOutputRoot, saveRunFiles, buildExcelBuffer, getDefaultYamlPath, getEnvYamlPath } from "./fileStore";
import { v4 as uuidv4 } from "uuid";
import { statSync, readFileSync } from "fs";
import { TX_CODE_ORDER } from "./constants";

// ── Types ─────────────────────────────────────────────────────────────────────

export type StepStatus = "ok" | "skipped" | "error" | "manual";

export interface StepOutcome {
  step: string;
  status: StepStatus;
  detail?: string;
}

export interface OrchestrationResult {
  runId: string;
  success: boolean;
  steps: StepOutcome[];
  btpClientId: string;
  muleFeatureBranch: string;
  prUrl: string;
  runResult?: RunResult;
  errors: string[];
}

// ── Orchestration state ───────────────────────────────────────────────────────

interface State {
  btpClientId: string;
  solaceEnumVersionId: string;
  solaceEventVersionId: string;
  muleFeatureBranch: string;
  yamlDiffSummary: string;
  translationRowCount: number;
}

function emptyState(): State {
  return {
    btpClientId: "",
    solaceEnumVersionId: "",
    solaceEventVersionId: "",
    muleFeatureBranch: "",
    yamlDiffSummary: "",
    translationRowCount: 0,
  };
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/** Prefer input field value, fall back to env var, then default. */
function cfg(inputVal: string | undefined, envKey: string, fallback = ""): string {
  return (inputVal && inputVal.trim()) ? inputVal.trim() : (process.env[envKey] ?? fallback);
}

async function safeRun<T>(
  stepName: string,
  fn: () => Promise<T>,
  outcomes: StepOutcome[]
): Promise<T | null> {
  try {
    const result = await fn();
    outcomes.push({ step: stepName, status: "ok" });
    return result;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    const isTodo = msg.includes("not yet implemented") || msg.includes("not implemented");
    outcomes.push({
      step: stepName,
      status: isTodo ? "skipped" : "error",
      detail: msg.split("\n")[0],
    });
    return null;
  }
}

function readYamlSafe(filePath: string): string {
  const stats = statSync(filePath);
  if (stats.size > 10 * 1024 * 1024) throw new Error("Source YAML file too large");
  return readFileSync(filePath, "utf-8");
}

// ── Main orchestrator ─────────────────────────────────────────────────────────

export async function orchestrate(input: OnboardingInput): Promise<OrchestrationResult> {
  const runId = uuidv4();
  const outcomes: StepOutcome[] = [];
  const errors: string[] = [];
  const state = emptyState();

  // Derive integration config: user-supplied fields take priority over env vars
  const btp   = input.btp;
  const solace = input.solace;
  const gh     = input.github;

  // ── Step 1: BTP Application Creation ─────────────────────────────────────
  const step1Result = await safeRun("Step 1 — BTP App Creation", () =>
    runStep1({
      apiHost:        cfg(btp?.apiHost,        "BTP_API_HOST"),
      tokenUrl:       cfg(btp?.tokenUrl,       "BTP_TOKEN_URL"),
      clientId:       cfg(btp?.clientId,       "BTP_CLIENT_ID"),
      clientSecret:   cfg(btp?.clientSecret,   "BTP_CLIENT_SECRET"),
      productName:    cfg(btp?.productName,    "BTP_PRODUCT_NAME"),
      quotaPlan:      cfg(btp?.quotaPlan,      "BTP_QUOTA_PLAN"),
      d3ApproverEmail: cfg(btp?.d3ApproverEmail, "BTP_D3_APPROVER_EMAIL"),
      appName:        `3pl-${input.translation?.receiver_name ?? input.country_key}`,
      appDescription: `3PL onboarding for ${input.country_key}`,
      partnerId:      input.translation?.receiver_name ?? "",
      countryCode:    input.country_key,
    }), outcomes
  );
  if (step1Result) state.btpClientId = step1Result.clientId;

  // ── Step 2: BTP Value Mapping ─────────────────────────────────────────────
  await safeRun("Step 2 — BTP Value Mapping", () =>
    runStep2({
      baseUrl:           cfg(btp?.suiteBaseUrl,         "BTP_SUITE_BASE_URL", "https://effem-glb-ci-dev01-pr.integrationsuite.cfapps.us21.hana.ondemand.com"),
      tokenUrl:          cfg(btp?.tokenUrl,              "BTP_TOKEN_URL"),
      clientId:          cfg(btp?.clientId,              "BTP_CLIENT_ID"),
      clientSecret:      cfg(btp?.clientSecret,          "BTP_CLIENT_SECRET"),
      artifactId:        cfg(btp?.valuemapArtifactId,    "BTP_VALUEMAP_ARTIFACT_ID"),
      packageId:         cfg(btp?.valuemapPackageId,     "BTP_VALUEMAP_PACKAGE_ID"),
      partnerId:         state.btpClientId || (input.translation?.receiver_name ?? ""),
      countryCode:       input.country_key,
      mappingEntries:    [],
    }), outcomes
  );

  // ── Step 3: Solace Event Portal ───────────────────────────────────────────
  const step3Result = await safeRun("Step 3 — Solace Event Portal", () =>
    runStep3({
      apiBase:         "https://api.solace.dev",
      apiToken:        cfg(solace?.epApiToken,      "SOLACE_EP_API_TOKEN"),
      enumId:          cfg(solace?.enumId,           "SOLACE_ENUM_ID"),
      eventId:         cfg(solace?.eventId,          "SOLACE_EVENT_ID"),
      producerAppId:   cfg(solace?.producerAppId,    "SOLACE_PRODUCER_APP_ID"),
      producerAppName: cfg(solace?.producerAppName,  "SOLACE_PRODUCER_APP_NAME"),
      consumerAppId:   cfg(solace?.consumerAppId,    "SOLACE_CONSUMER_APP_ID"),
      consumerAppName: cfg(solace?.consumerAppName,  "SOLACE_CONSUMER_APP_NAME"),
      environmentId:   cfg(solace?.environmentId,   "SOLACE_ENVIRONMENT_ID"),
      brokerId:        cfg(solace?.brokerId,         "SOLACE_BROKER_ID"),
      countryCode:     input.country_key,
      regionIso:       solace?.regionIso ?? "",
      partnerId:       state.btpClientId || (input.translation?.receiver_name ?? ""),
      subscriptionTopics: solace?.subscriptionTopics ?? [],
    }), outcomes
  );
  if (step3Result) {
    state.solaceEnumVersionId = step3Result.enumVersionId;
    state.solaceEventVersionId = step3Result.eventVersionId;
  }

  // ── Step 4: Solace Queue Subscription Patch ───────────────────────────────
  await safeRun("Step 4 — Solace Queue Patch", () =>
    runStep4({
      sempHost:     cfg(solace?.sempHost,     "SOLACE_SEMP_HOST"),
      sempPort:     cfg(solace?.sempPort,     "SOLACE_SEMP_PORT", "943"),
      sempUsername: cfg(solace?.sempUsername, "SOLACE_SEMP_USERNAME"),
      sempPassword: cfg(solace?.sempPassword, "SOLACE_SEMP_PASSWORD"),
      vpnNameUat:   cfg(solace?.vpnNameUat,   "SOLACE_VPN_UAT"),
      vpnNameProd:  cfg(solace?.vpnNameProd,  "SOLACE_VPN_PROD"),
      queueName:    cfg(solace?.queueName,    "SOLACE_QUEUE_NAME"),
      newTopics:    solace?.subscriptionTopics ?? [],
      partnerId:    state.btpClientId || (input.translation?.receiver_name ?? ""),
    }), outcomes
  );

  // ── Step 5: Solace Git / input.json ──────────────────────────────────────
  await safeRun("Step 5 — Solace Git / input.json", () =>
    runStep5({
      githubApiBase:      "https://api.github.com",
      repoOwner:          cfg(gh?.solaceRepoOwner,     "SOLACE_GITHUB_REPO_OWNER"),
      repoName:           cfg(gh?.solaceRepoName,      "SOLACE_GITHUB_REPO_NAME"),
      baseBranch:         cfg(gh?.solaceBaseBranch,    "SOLACE_GITHUB_BASE_BRANCH", "dev"),
      githubPat:          cfg(gh?.githubPat,           "GITHUB_PAT"),
      inputJsonPath:      cfg(gh?.solaceInputJsonPath,  "SOLACE_INPUT_JSON_PATH"),
      countryCode:        input.country_key,
      partnerId:          state.btpClientId || (input.translation?.receiver_name ?? ""),
      subscriptionTopics: solace?.subscriptionTopics ?? [],
      enumVersionId:      state.solaceEnumVersionId,
      eventVersionId:     state.solaceEventVersionId,
    }), outcomes
  );

  // ── Step 6: MuleSoft Feature Branch ──────────────────────────────────────
  const step6Result = await safeRun("Step 6 — Mule Feature Branch", () =>
    runStep6({
      githubApiBase: "https://api.github.com",
      repoOwner:     cfg(gh?.muleRepoOwner,   "MULE_GITHUB_REPO_OWNER"),
      repoName:      cfg(gh?.muleRepoName,    "MULE_GITHUB_REPO_NAME"),
      baseBranch:    cfg(gh?.muleBaseBranch,  "MULE_GITHUB_BASE_BRANCH", "dev"),
      githubPat:     cfg(gh?.githubPat,       "GITHUB_PAT"),
      branchPrefix:  "feature/3pl-onboard",
      partnerId:     state.btpClientId || (input.translation?.receiver_name ?? ""),
      countryCode:   input.country_key,
    }), outcomes
  );
  if (step6Result) state.muleFeatureBranch = step6Result.branchName;

  // ── Steps 7 + 8: YAML Patch + Translation Rows ───────────────────────────
  let runResult: RunResult | undefined;
  try {
    const devBlock  = buildYamlBlock(input);
    const tstBlock  = input.nav_tst  ? buildYamlBlockForEnv(input, input.nav_tst)  : null;
    const prodBlock = input.nav_prod ? buildYamlBlockForEnv(input, input.nav_prod) : null;

    // Patch dev.yaml
    const devYamlPath = getDefaultYamlPath();
    let patchedDev = patchYaml(readYamlSafe(devYamlPath), input.country_key, devBlock);

    // Patch app.yaml (3-pass)
    let patchedApp: string | undefined;
    const appYamlPath = getEnvYamlPath("app");
    if (appYamlPath) {
      let yaml = readYamlSafe(appYamlPath);
      yaml = patchValidCountriesList(yaml, input.country_code);
      yaml = patchNavisionInstance(yaml, input.country_code, input.country_key);
      patchedApp = patchYaml(yaml, input.country_key, devBlock);
    }

    // Patch tst.yaml
    let patchedTst: string | undefined;
    const tstYamlPath = getEnvYamlPath("tst");
    if (tstBlock && tstYamlPath) {
      patchedTst = patchYaml(readYamlSafe(tstYamlPath), input.country_key, tstBlock);
    }

    // Patch prod.yaml
    let patchedProd: string | undefined;
    const prodYamlPath = getEnvYamlPath("prod");
    if (prodBlock && prodYamlPath) {
      patchedProd = patchYaml(readYamlSafe(prodYamlPath), input.country_key, prodBlock);
    }

    // Translation rows + Excel/CSV
    const rows        = buildTranslationRows(input);
    const excelBuffer = await buildExcelBuffer(rows);
    const csvContent  = buildCsv(rows);

    const txEnabled = TX_CODE_ORDER.filter(c => input.nav.transaction_types[c]?.enabled);

    runResult = {
      run_id:           runId,
      country_key:      input.country_key,
      country_code:     input.country_code,
      nav_host:         input.nav.host,
      nav_company:      input.nav.company,
      routing_code:     input.nav.routing_code,
      tx_enabled:       txEnabled,
      receiver_name:    input.translation.receiver_name,
      yaml_block:       devBlock,
      ...(patchedApp  && { app_yaml_block:  devBlock }),
      ...(tstBlock     && { tst_yaml_block:  tstBlock }),
      ...(prodBlock    && { prod_yaml_block: prodBlock }),
      translation_rows: rows,
      created_at:       new Date().toISOString(),
    };

    state.yamlDiffSummary     = devBlock;
    state.translationRowCount = rows.length;

    ensureOutputRoot();
    saveRunFiles(runId, runResult, patchedDev, excelBuffer as Buffer, csvContent, patchedApp, patchedTst, patchedProd);

    outcomes.push({ step: "Step 7 — Mule YAML Patch",   status: "ok" });
    outcomes.push({ step: "Step 8 — Translation Rows",  status: "ok", detail: `${rows.length} rows` });
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    errors.push(msg);
    outcomes.push({ step: "Step 7 — Mule YAML Patch",  status: "error", detail: msg });
    outcomes.push({ step: "Step 8 — Translation Rows", status: "error" });
  }

  // ── Step 9: MuleSoft PR + Notification ───────────────────────────────────
  let prUrl = "";
  const step9Result = await safeRun("Step 9 — Mule PR + Notify", () =>
    runStep9({
      githubApiBase:        "https://api.github.com",
      repoOwner:            cfg(gh?.muleRepoOwner,        "MULE_GITHUB_REPO_OWNER"),
      repoName:             cfg(gh?.muleRepoName,         "MULE_GITHUB_REPO_NAME"),
      githubPat:            cfg(gh?.githubPat,            "GITHUB_PAT"),
      baseBranch:           cfg(gh?.muleBaseBranch,       "MULE_GITHUB_BASE_BRANCH", "dev"),
      reviewerGithubHandle: cfg(gh?.muleReviewerHandle,   "MULE_REVIEWER_GITHUB_HANDLE"),
      prLabel:              "3pl-onboarding",
      notificationChannel:  (cfg(gh?.notificationChannel, "NOTIFICATION_CHANNEL", "email")) as "email" | "teams" | "slack",
      reviewerEmail:        cfg(gh?.reviewerEmail,        "REVIEWER_EMAIL"),
      smtpHost:             cfg(gh?.smtpHost,             "SMTP_HOST"),
      smtpFrom:             cfg(gh?.smtpFrom,             "SMTP_FROM"),
      teamsWebhookUrl:      cfg(gh?.teamsWebhookUrl,      "TEAMS_WEBHOOK_URL"),
      slackWebhookUrl:      cfg(gh?.slackWebhookUrl,      "SLACK_WEBHOOK_URL"),
      featureBranch:        state.muleFeatureBranch,
      countryKey:           input.country_key,
      partnerId:            state.btpClientId || (input.translation?.receiver_name ?? ""),
      yamlDiffSummary:      state.yamlDiffSummary,
      translationRowCount:  state.translationRowCount,
    }), outcomes
  );
  if (step9Result) prUrl = step9Result.prUrl;

  // ── Step 10: Manual gate ──────────────────────────────────────────────────
  outcomes.push({
    step:   "Step 10 — SharePoint / Manual Review",
    status: "manual",
    detail: "Upload translation XLSX to SharePoint TranslationData_TST list",
  });

  const overallSuccess =
    outcomes.every(o => o.status !== "error") && errors.length === 0;

  return {
    runId,
    success: overallSuccess,
    steps:   outcomes,
    btpClientId:       state.btpClientId,
    muleFeatureBranch: state.muleFeatureBranch,
    prUrl,
    runResult,
    errors,
  };
}
